url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/Utils/ConsoleWriter.java

```diff
@@ -0,0 +1,226 @@
+package Utils;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+
+public class ConsoleWriter {
+	public static String consoleID = "";
+	private String name = "";
+	Semaphore stateChange = new Semaphore(1, true);// binary semaphore, fair
+	private ConsoleWriterThread consoleWriterThread;
+
+	// The backing data structure will be blocking queue
+	private final Queue<String> messageQueue;
+
+	public ConsoleWriter() {
+		this(null);
+	}
+
+	public ConsoleWriter(String name) {
+		this.name = name;
+		messageQueue = new LinkedBlockingQueue<String>();
+	}
+
+	/**
+	 * This should be called whenever state has changed that might cause the console writer to do something.
+	 */
+	protected void stateChanged() {
+		stateChange.release();
+	}
+
+	protected void print(String msg) {
+		print(msg, null);
+	}
+
+	/** Print message with exception stack trace */
+	protected void print(String msg, Throwable e) {
+		StringBuffer sb = new StringBuffer();
+		sb.append("[ConsoleWriter]");
+		sb.append(": ");
+		sb.append(msg);
+		sb.append("\n");
+		if (e != null) {
+			sb.append(StringUtil.stackTraceString(e));
+		}
+
+		System.out.print(sb.toString());
+	}
+
+	/**
+	 * Console writer scheduler thread, calls sendingMessage() whenever a state change has been signaled.
+	 */
+	private class ConsoleWriterThread extends Thread {
+		private volatile boolean goOn = false;
+
+		private ConsoleWriterThread(String name) {
+			super(name);
+		}
+
+		@Override
+		public void run() {
+			goOn = true;
+
+			while (goOn) {
+				try {
+					stateChange.acquire();
+					while (sendingMessage()) {
+						;
+					}
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)